### ISSUE TITLE
[Bug] File extension patterns not detected when plugin is required explicitly using require() (due to PnP)

### ISSUE THREAD

**Issue by Waghabond** [prettier/prettier#13276]
## Steps to Reproduce:
Steps provided in repo readme: [Prettier Plugin Explicit Require Bug demo repo](https://github.com/Waghabond/prettier-plugin-explicit-require-bug)

**Environments:**

- Prettier Version: 2.7.1
- Running Prettier via: CLI (Yarn PnP)
- Runtime: Node.js v16.13.2
- Operating System: WSL2 - Ubuntu 20.04
- Prettier plugins (if any): @prettier/plugin-php, @shufo/prettier-plugin-blade, prettier-plugin-sh (and probably all other plugins)

## Description:

When using PnP plugins are not being loaded as yet (see issue #8474 for that discussion) . A work around for this is using require() to add them explicitly in a .prettierrc.js (this mostly works well). 
However, plugins that are required this way dont seem to seem to have their file extension glob patterns detected by prettier, they're loaded and work correctly if the correct files are passed in eg. with a command like `yarn prettier -w *.php` .
Prettier won't automatically detect the files when it recursively traverses a directory eg. with a command like `yarn prettier -w .`  (.php files wont be detected with this).

  ## Why this problem arises and Solution:
After some digging, I found the reason why i'm facing sunch a weird issue: 
> Doing it this way, the glob pattern resolving stuff will not detect file extensions declared by the plugins. This is because the context object (which resolves the glob into filepaths) is loaded before the .prettierrc file (see #13276). As a result the plugins get loaded due to the configuration in the prettierrc but the files they're supposed to format won't get detected. 
>
> This makes it so that there is no option but to pass the plugins using the `--plugin` cli argument. 
> I think the only sensible way to handle the issue is what has been previously sugested in this thread: 
> - in the `cli.js` file we should introspect the `package.json` file the `.prettierrc` file for plugin names 
> - Then use `require()` to load them.  
>
> It's important that the plugin loading step is done in `cli.js` before the context is loaded because otherwise glob resolution doesn't work correctly
>
> _Originally posted by @Waghabond in https://github.com/prettier/prettier/issues/8474#issuecomment-1213386876_

A solution proposed in https://github.com/prettier/prettier/issues/8474 would fix this problem.

Prettier's `cli.js` should have an aditional step wherein it reads the `package.json` and `.prettierrc` to find any installed prettier modules and then it should try and require them using `require()`
The changes would not be breaking as it would be an additional step, and the normal node_module based resolutions can continue happening as they were.

 I have very little understanding of how js modules work but I'm willing to draft up a pull request but if there is any interest for this!

## TLDR; Hacky fix for people who are currently using Yarn 2 with PnP:
1. use a `prettierrc.js` file with the plugin option and explicitly require your plugins like so:
```js
module.exports = {
    // Whatever plugins you need here:
    plugins: [require.resolve("@prettier/plugin-php"), require.resolve("prettier-plugin-sh")],
}
```
2. To perform formatting on your project from the cli, you will need this command:
```bash
# one --plugin for each of your plugins
yarn prettier -w ./ --plugin="@prettier/plugin-php" --plugin="prettier-plugin-sh"
```
3. Convenience tip add this to `package.json` scripts and you will be able to run it using `yarn format`:
```json
"scripts": {
    "format": "prettier -w ./ --plugin=@prettier/plugin-php --plugin=prettier-plugin-sh"
}
```
4. If using vscode, you will need to additionally add the following into your vscode settings, and `.prettierrc.js` respectively.
  - ```json 
    "prettier.documentSelectors": ["**/*.php"]
    ```
  - ```js
    module.exports = {
        // whatever other stuff...
        overrides: [
            {
                files: ["*.php"],
                options: {
                    parser: "php",  // vscode will not be able to infer the parser for some reason so we need to explicitly specify it.
                },
            },
        ]
    }

**Comment by Waghabond** [comment]
Bumping this, is there any updates?

**Comment by Techn1x** [comment]
It looks like everyone in this thread is also experiencing the same issue, but with pnpm
https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/38

That plugin is meant to be registered for gts/gjs extensions, but seems to not be registering for those on pnpm. They found the same as in your opening post - that adding the `"prettier.documentSelectors"` line to vscode, as well as explicitly adding the overrides to their prettierrc.js file was a workaround

**Comment by Techn1x** [comment]
Since prettier v3 has dropped support for plugin search https://github.com/prettier/prettier/pull/14759 everyone's going to need to rely on explicitly listing plugins in this array, we should probably aim to fix this for v3...... cc @fisker ?

**Comment by Waghabond** [comment]
@jorismak does the "Hacky Fix" i show above not work for you?

**Comment by thorn0** [comment]
@jorismak check v3.1 please. If not fixed, let's reopen

**Comment by jorismak** [comment]
👍
